### PR TITLE
DAOS-7300 test: Target auto eviction (checksum faults)

### PR DESCRIPTION
**Pull request by rpadma2**
Skip-unit-tests: true
Skip-nlt: true
Test-tag: pr scrubber

Summary:
-  The test set the scrub-thresh value to a certain limit. Once the total checksum corrupt count reaches scrub-thresh value, the target should be evicted automatically by the scrubber.
- The test compares the active targets before and after the checksum corruption. The active targets should be less at the end of the test case.

Signed-off-by: rpadma2 <ravindran.padmanabhan@intel.com>

### CODE REVIEW

**Comment by rpadma2** [comment]
NOTE: I need to add the faults avocado tag since the test uses fault injection framework. Will do it after I have good test run.

**Comment by daltonbohning** [comment]
> Test-tag-hw-medium: pr,hw,medium,ib2 scrubber

FYI: CI will figure out where the tests need to run, so you can just do this:
`Test-tag: pr scrubber`

File: src/tests/ftest/scrubber/target_auto_eviction.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    """Inject Checksum Fault with scrubber enabled
+    and scrubber threshold set to a certain value.
+
+    :avocado: recursive
+    """
+    def test_scrubber_ssd_auto_eviction(self):
+        """JIRA ID: DAOS-7300
+
+            1. Create checksum faults above scrubber threshold
+            and see whether SSD auto eviction works as expected.
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/target_auto_eviction.py:23]: Minor format
```suggestion
        """JIRA ID: DAOS-7300

        1. Create checksum faults above scrubber threshold
        and see whether SSD auto eviction works as expected.
```

File: src/tests/ftest/scrubber/target_auto_eviction.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        self.ior_cmd.flags.update(flags[0], "ior.flags")
+        self.ior_cmd.dfs_oclass.update(obj_class[0])
+        self.ior_cmd.dfs_dir_oclass.update(obj_class[0])
+        self.create_pool_cont_with_scrubber(pool_prop=pool_prop, cont_prop=cont_prop)
+        self.dmg_cmd.pool_query(self.pool.uuid)
+        initial_metrics = self.scrubber.get_scrub_corrupt_metrics()
+        for test in transfer_block_size:
+            self.ior_cmd.transfer_size.update(test[0])
+            self.ior_cmd.block_size.update(test[1])
+        self.run_ior_and_check_scruber_status(pool=self.pool, cont=self.container)
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/target_auto_eviction.py:49]: This is only going to use the last `transfer_block_size`

File: src/tests/ftest/scrubber/target_auto_eviction.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        time.sleep(60)
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/target_auto_eviction.py:50]: Can we please add a comment about why we're sleeping and why for this long?

File: src/tests/ftest/scrubber/target_auto_eviction.yaml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+container:
+  type: POSIX
+  control_method: daos
+  oclass: RP_2G1
+  properties: "cksum:crc16"
+ior:
+  ior_timeout: 60
+  client_processes:
+    np_16:
+        np: 1
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/target_auto_eviction.yaml:69]:
```suggestion
  client_processes:
    np: 1
```

File: src/tests/ftest/scrubber/target_auto_eviction.yaml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  iorflags:
+    ior_flags:
+      - "-v -W -w -r -R"
+    ior_api:
+      - DFS
+    transfer_block_size:
+      - [256B, 2M]
+      - [1M, 32M]
+    obj_class:
+      - RP_2G1
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/target_auto_eviction.yaml:82]: Based on test usage, we can simplify this and remove the need to set manually in the `.py`
```suggestion
    flags: "-v -W -w -r -R"
    api: DFS
    transfer_size: 1M
    block_size: 32M
    dfs_oclass: RP_2G1
    dfs_dir_oclass: RP_2G1
```

**Review by daltonbohning — CHANGES_REQUESTED** [review]

File: src/tests/ftest/scrubber/target_auto_eviction.py (hunk 1 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.py:23]: Done.

File: src/tests/ftest/scrubber/target_auto_eviction.py (hunk 3 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.py:50]: Added comment. We need to wait for the scrubber to take action of identifying the scrubber threshold value and evicting the target,

File: src/tests/ftest/scrubber/target_auto_eviction.yaml (hunk 4 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.yaml:69]: Done.

File: src/tests/ftest/scrubber/target_auto_eviction.yaml (hunk 5 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.yaml:82]: Good input. Done. Thanks.

**Comment by rpadma2** [comment]
I will add the reviewers later... I need to skip this test due to a JIRA ticket: https://daosio.atlassian.net/browse/DAOS-10377. Presently, want to make sure the test runs as expected with the script changes based on code review.

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/scrubber/target_auto_eviction.yaml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  transport_config:
+    allow_insecure: True
+dmg:
+  transport_config:
+    allow_insecure: True
+pool:
+  mode: 146
+  name: daos_server
+  scm_size: 6000000000
+  nvme_size: 54000000000
```
Comment (shimizukko) [inline src/tests/ftest/scrubber/target_auto_eviction.yaml:54]:
```suggestion
  scm_size: 6G
  nvme_size: 54G
```
I think our standard now is human-readable.

File: src/tests/ftest/scrubber/target_auto_eviction.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        initial_metrics = {}
+        final_metrics = {}
+        self.create_pool_cont_with_scrubber(pool_prop=pool_prop, cont_prop=cont_prop)
+        self.dmg_cmd.pool_query(self.pool.uuid)
+        initial_metrics = self.scrubber.get_scrub_corrupt_metrics()
+        self.run_ior_and_check_scruber_status(pool=self.pool, cont=self.container)
+        # Wait for a minute for the scrubber to take action and evict target
+        # after corruption threshold reached.
+        time.sleep(60)
+        self.dmg_cmd.pool_query(self.pool.uuid)
```
Comment (shimizukko) [inline src/tests/ftest/scrubber/target_auto_eviction.py:41]: Is this pool query for debugging purpose?

File: src/tests/ftest/scrubber/target_auto_eviction.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/python
+"""
+  (C) Copyright 2018-2021 Intel Corporation.
```
Comment (shimizukko) [inline src/tests/ftest/scrubber/target_auto_eviction.py:3]:
```suggestion
  (C) Copyright 2018-2022 Intel Corporation.
```

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/scrubber/target_auto_eviction.yaml (hunk 6 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.yaml:54]: Sure. Will do. Can do the changes along with the skipForTicket pragma changes.

File: src/tests/ftest/scrubber/target_auto_eviction.py (hunk 7 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.py:41]: Yes. The target information should change based on the scrubber eviction.

File: src/tests/ftest/scrubber/target_auto_eviction.py (hunk 8 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/target_auto_eviction.py:3]: Will do.

**Comment by rpadma2** [comment]
DaosCore/OfflineDrain test failures seems to related to node issues. Servers didn't boot properly due to ib issue on certain nodes.

Scrubber/csum failure should be related to DAOS-9318. intermitent failure. I will be investigating the ticket later.

**Comment by rpadma2** [comment]
@ryon-jensen : This PR is ready for merge with the feature/csum-scrubbing branch. The test is presently skipped due to DAOS-10377. Please merge the PR to the feature branch when you get time . Thanks.

**Pull request merged**